lib/plurimath/omml/parse.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+require "parslet"
+module Plurimath
+  class Omml
+    class Parse < Parslet::Parser
+      rule(:sequence)  { (iteration >> iteration) | iteration }
+      rule(:iteration) { (expression >> expression) | expression | number_text }
+
+      rule(:open_sub_sup)  { parse_tag(:open, Constants::SUB_SUP_TAG) }
+      rule(:close_sub_sup) { parse_tag(:close, Constants::SUB_SUP_TAG) }
+
+      rule(:attributes) do
+        (attribute.as(:name) >> str("=") >> quoted_string).repeat
+      end
+
+      rule(:sub_sup_tags) do
+        (open_sub_sup >> tag.as(:fonts) >> sub_sup_values >> close_sub_sup)
+      end
+
+      rule(:attribute) do
+        (match["a-zA-Z"].repeat >> str(":") >> match["a-zA-Z0-9"].repeat) |
+          match["a-zA-Z0-9"].repeat
+      end
+
+      rule(:number_text) do
+        match["0-9"].repeat(1).as(:number) |
+          match["a-zA-Z"].as(:text).repeat |
+          match("[^a-zA-Z]")
+      end
+
+      rule(:tag) do
+        omission_tag |
+          (parse_tag(:open) >> sequence.as(:iteration) >> parse_tag(:close)) |
+          (parse_tag(:open) >> parse_tag(:close))
+      end
+
+      rule(:sub_sup_values) do
+        (tag.as(:first_value) >> tag.as(:second_value) >> tag.as(:third_value)) |
+          (tag.as(:first_value) >> tag.as(:second_value)) |
+          tag.as(:first_value)
+      end
+
+      rule(:expression) do
+        sub_sup_tags |
+          (tag >> sequence.as(:sequence)) |
+          tag
+      end
+
+      root :expression
+
+      def array_to_expression(array, name = nil)
+        initial_type = array.first.class
+        array.reduce do |expr, tag|
+          expr = str_to_expression(expr, name) if expr.is_a?(initial_type)
+          expr | str_to_expression(tag, name)
+        end
+      end
+
+      def str_to_expression(string, name)
+        return str(string) if name.nil?
+
+        str(string).as(name)
+      end
+
+      def quoted_string
+        (str('"') >> match("[^\"]").repeat.as(:value) >> str('"')) |
+          (str("'") >> match("[^\']").repeat.as(:value) >> str("'"))
+      end
+
+      def omission_tag
+        tag = str("<w:") | str("<m:")
```

Comment:
The direct usage of `<w:` and `<m:` for parsing is very dangerous, because these are actually XML namespace aliases. The actual definitions of the xmlns (XML namespace) are like this:
https://github.com/transpect/docx2hub/blob/519ad7c22869d5b1536809ab713ac64951cff28d/xsl/omml2mml/omml2mml.xsl#L12-L14

i.e
```xml
  xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" 
  xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" 
```

For example, if the xmlns line looked like this:
```
  xmlns:moo="http://schemas.openxmlformats.org/wordprocessingml/2006/main" 
```

Then the tags will begin like `<moo:oMath ...`.

I would strongly suggest that we use an XML parser like Ox (the native Ruby XML parser, over Nokogiri), to parse this XML.